frontend/ui/trapper.lua

```diff
@@ -0,0 +1,133 @@
+--[[--
+Trapper module: provides methods for simple interaction with UI, to be
+used by linear jobs between their steps.
+
+Allows code to trap UI (give progress info to UI, ask for user choice),
+or get trapped by UI (get interrupted).
+Mostly done with coroutines, but hides their usage for simplicity.
+]]
+
+-- TODO: move generic parts of ui/wikipedia.lua Wikipedia:createEpubWithUI() here
+-- ie:
+-- function Trapper:uiFeedback(text, confirmbox_else_infomessage, ok_text, cancel_text)
+--   return go_on / choice
+-- + some coroutine bootstrap methods to abstract all that from callers
+
+local UIManager = require("ui/uimanager")
+local ffiutil = require("ffi/util")
+local logger = require("logger")
+
+local Trapper = {}
+
+--- Allow function code to be trapped
+--
+-- Simply wrap function into a coroutine, which is a prerequisite
+-- for all our methods (this simply abstracts the coroutine
+-- business to our callers)
+-- The function should be the last step of some event processing code,
+-- as we may return early (with coroutine.yield()), and later be
+-- resumed by UIManager. So any following (unwrapped) code would
+-- be then executed while func is half-done, with unintended consequences.
+--
+-- @func function function that can use us
```

Comment:
Did you try this with LDoc? I'd think you'd have to use `@param func Some description`

Reply:
Yes, I think something like this produces what you were going for.

```lua
--- Allows function code to be trapped.
--
-- Simple wrapper function for a coroutine, which is a prerequisite
-- for all our methods (this simply abstracts the @{coroutine}
-- business to our callers)
--
-- The function should be the last step of some event processing code,
-- as we may return early (with @{coroutine.yield|coroutine.yield()}), and later be
-- resumed by @{ui.uimanager|UIManager}. So any following (unwrapped) code would
-- be then executed while `func` is half-done, with unintended consequences.
--
-- @param function function reference to function to wrap
function Trapper:wrap(func)
    local co = coroutine.create(func)
    return coroutine.resume(co)
end
```